Use a virtual threads friendly pool with Jackson

mariofusco opened the pull request:
This pull request is a follow up of the work done to make the [internal pool used by Jackson configurable and pluggable](https://github.com/FasterXML/jackson-core/pull/1064). It is not intended to be merged until Jackson 2.16 won't be released. Also this is the porting of the [same pull request sent to quarkus](https://github.com/quarkusio/quarkus/pull/35816).

vietj commented:
merge is on hold until 2.16.0 is released

franz1981 commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,208 @@
+
+  static class VirtualPoolHolder {
+    // Lazy on-demand initialization
+    private static final RecyclerPool<BufferRecycler> virtualPool = new StripedLockFreePool(4);
```
use the Runtime.getRuntime().availableProcessors here

mariofusco replied:
Done, thanks for reminder.

vietj commented:
@mariofusco @franz1981 we will release 4.5.0 this week, I'm assuming we cannot yet merge this PR ?

mariofusco commented:
> @mariofusco @franz1981 we will release 4.5.0 this week, I'm assuming we cannot yet merge this PR ?

The final release of Jackson 2.16 is not out yet ( see https://central.sonatype.com/artifact/com.fasterxml.jackson.core/jackson-core/versions ) so, unless it is ok for you to depend on that 2.16.0-rc1, I'm afraid that this PR cannot be merged yet.

vietj commented:
I will move this issue to 4.5.1

On Wed, Nov 15, 2023 at 11:07 AM Mario Fusco ***@***.***>
wrote:

> @mariofusco <https://github.com/mariofusco> @franz1981
> <https://github.com/franz1981> we will release 4.5.0 this week, I'm
> assuming we cannot yet merge this PR ?
>
> The final release of Jackson 2.16 is not out yet ( see
> https://central.sonatype.com/artifact/com.fasterxml.jackson.core/jackson-core/versions
> ) so, unless it is ok for you to depend on that 2.16.0-rc1, I'm afraid that
> this PR cannot be merged yet.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-vertx/vert.x/pull/4920#issuecomment-1812152061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCUB72QKRBTKEXAMYQ3YESH6TAVCNFSM6AAAAAA6RO4ZCSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMJSGE2TEMBWGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

magicprinc commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,208 @@
+        };
+      }
+
+      return new Predicate<Thread>() {
```
why not simply: `return thread->false;`

mariofusco replied:
We decided to avoid using lambdas here for performance reasons.

magicprinc commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,208 @@
+ */
+public class HybridJacksonPool implements RecyclerPool<BufferRecycler> {
+
+  static final RecyclerPool INSTANCE = new HybridJacksonPool();
```
Any chance to make it public? 🙏

mariofusco replied:
I made the field private and added a public `getInstance()` method.

magicprinc commented:
As you know, Jackson 2.16.0 is released
https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-core/2.16.0

mariofusco commented:
@vietj I think that this is ready to be merged now. I bumped the jackson version with this other PR https://github.com/vert-x3/vertx-dependencies/pull/157

franz1981 commented:
@mariofusco Please add some tests to be sure that it works as expected; by mocking the thread id(s) and verifying it hits/distribute the requests and/or reuse the same pool, if expected

franz1981 reviewed (changes requested):
@mariofusco Please add some tests to be sure that it works as expected (including not getting any weird array index out of bound exception); by mocking the thread id(s) and verifying it hits/distribute the requests and/or reuse the same pool, if expected.

Additionally, add something (I would use a counter into the node(s)) to report the current size of the pool; 
this type of telemetry will help to troubleshoot errors.

Decide if it makes sense to add a fastthreadlocal-only version to replace the current existing one too, and give vertx users a boost without relying on the existing implementation from Jackson (which uses SoftRef too, bleah!)

franz1981 commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,212 @@
+
+    private final XorShiftThreadProbe threadProbe;
+
+    private final AtomicReferenceArray<Node> heads;
```
replace with `topStack`

franz1981 commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,212 @@
+
+      Node next = heads.get(vThreadBufferRecycler.slot);
+      while (true) {
+        if (heads.compareAndSet(vThreadBufferRecycler.slot, next, newHead)) {
```
right before this: `newHead.level = next.level + 1` - level doesn't need to be volatile, bacause is correctly acquired/released by reading the `Node`'s instance and performing this `compareAndset`.

franz1981 commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,212 @@
+
+    private static class Node {
+      final VThreadBufferRecycler value;
+      Node next;
```
Add `int level = 0` field

franz1981 commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,212 @@
+      int mask = (size - 1) << CACHE_LINE_SHIFT;
+      this.threadProbe = new XorShiftThreadProbe(mask);
+    }
+
```
add a method `int size` which round on all the topStacks and sum `topStack.ge().level` for each

mariofusco commented:
@franz1981 I implemented what you suggested and added a couple of simple tests. Please let me know if there is something more that I could check or improve.

mariofusco commented:
@vietj @franz1981 is this ok to be merged?

franz1981 commented on src/test/java/io/vertx/core/json/jackson/HybridJacksonPoolTest.java:
```diff
@@ -0,0 +1,68 @@
+
+    for (int i = 0; i < nThreads; i++) {
+      int threadIndex = i;
+      new Thread(() -> {
```
This cat kill the CI, let's find an alternative way to achieve the same (expose a method which use the Id and always hit the striped variant?)

franz1981 replied:
Or assume virtual threads are available and use them via reflection?

franz1981 commented on src/test/java/io/vertx/core/json/jackson/HybridJacksonPoolTest.java:
```diff
@@ -0,0 +1,68 @@
+    // check that all resources have been released back to the pool
+    assertEquals(nThreads, virtualPool.size());
+
+    int avgResourcesNrPerStripe = nThreads / stripesCount;
```
I love it, nice!!!

franz1981 commented on src/test/java/io/vertx/core/json/jackson/HybridJacksonPoolTest.java:
```diff
@@ -26,6 +29,11 @@ public void testVirtualThreadPoolWithSingleThread() {
 
   @Test
   public void testVirtualThreadPoolWithMultipleThreads() {
+    if (!VirtualThreadRunner.hasVirtualThread()) {
```
you can use assume for this

mariofusco commented:
@vietj @franz1981 I modified my test case to use virtual threads if they are available and completely skip the test otherwise. Please let me know if there is something else to be done here. If not I don't see any further reason to hold this pull request.

franz1981 commented on src/test/java/io/vertx/core/json/jackson/HybridJacksonPoolTest.java:
```diff
@@ -65,4 +73,29 @@ public void testVirtualThreadPoolWithMultipleThreads() {
     }
   }
+
+  private static class VirtualThreadRunner {
```
nice!

franz1981 reviewed (approved):
LGTM well done @mariofusco !

mariofusco commented:
Hi @vietj is there any other outstanding issue why you're holding this PR? (not rushing, just checking if I missed to do something)

vietj commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,238 @@
+ * {@link AtomicReferenceArray} where each head has a distance of 16 positions from the adjacent ones to prevent
+ * the false sharing problem.
+ */
+public class HybridJacksonPool implements RecyclerPool<BufferRecycler> {
```
I think we can make this package private instead of being public

mariofusco replied:
The idea was to use the same pool also in Quarkus, if I make it package private this will be no longer possible.

vietj replied:
ah ok this makes sense, then just add a slight comment that this class is reused elsewhere so we won't refactor it if that's not needed in order to preserve compatiblity

mariofusco replied:
Done.

vietj commented on src/main/java/io/vertx/core/json/jackson/HybridJacksonPool.java:
```diff
@@ -0,0 +1,238 @@
+    }
+  }
+
+  private static class XorShiftThreadProbe {
```
Some documentation is welcome :-)

vietj reviewed (changes requested):
A couple of minor improvements

mariofusco commented:
> A couple of minor improvements

@vietj I addressed one and replied to the other.

vietj commented:
@mariofusco can you squash everything in a single commit with a nice commit message :-) ?

vietj commented:
@mariofusco I see some activity on this PR, is it safe to merge ?

mariofusco commented:
> @mariofusco I see some activity on this PR, is it safe to merge ?

@vietj I squashed everything as you asked, but now I'm also testing the Quarkus integration as suggested by @franz1981 ( see https://github.com/quarkusio/quarkus/pull/38196 ). Please just give me a bit of time to check that it is also working as expected.

mariofusco commented:
@vietj @franz1981 I'm sorry, but I tested this integration and found a serious bug in `jackson-databind`, see https://github.com/quarkusio/quarkus/pull/38196#issuecomment-1892430695

I'm afraid that this pull request cannot be merged yet, till the problem won't be fixed in jackson and for sure we cannot use jackson 2.16.1 for this.

vietj commented:
thanks for raising this @mariofusco so for now it will be on hold

mariofusco commented:
@vietj @franz1981 @geoand I further investigated the problem, found the root cause and with my [last commit](https://github.com/eclipse-vertx/vert.x/pull/4920/commits/766f2326f3c3f7f85c657d3f8f9c2e2ba258946b) made my hybrid pool compatible with jackson-databind behaviour. I also test that with this fix all Quarkus integration tests in the jackson extension are green. 

I'm [still discussing](https://github.com/FasterXML/jackson-databind/issues/4321#issuecomment-1893504676) with @cowtowncoder if what jackson-databind does is correct or not, or if we can find a way to improve it. Anyway with my fix that discussion is now unrelated with this pool implementation and now I believe that it is totally safe to merge this as it is. 

@vietj if it is ok for you, please feel free merging this pull request.

mariofusco commented:
Hi @vietj, the problem that I found is now fixed in Jackson, but as I wrote in my former comment, with my last commit I found a solution that works even without that fix and I checked that this behaves correctly also on the quarkus integration side. Is there any outstanding work here or any other reason why you're still holding this pull request?

mariofusco commented:
> @mariofusco merged, thanks for the contribution, can you port this to master branch ?

@vietj see https://github.com/eclipse-vertx/vert.x/pull/5084

vietj commented:
Awesome thanks!

On Wed, Jan 31, 2024 at 10:36 AM Mario Fusco ***@***.***>
wrote:

> @mariofusco <https://github.com/mariofusco> merged, thanks for the
> contribution, can you port this to master branch ?
>
> @vietj <https://github.com/vietj> see #5084
> <https://github.com/eclipse-vertx/vert.x/pull/5084>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-vertx/vert.x/pull/4920#issuecomment-1918728771>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCQEAGEP7LKQLWMGTYDYRIGB3AVCNFSM6AAAAAA6RO4ZCSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJYG4ZDQNZXGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
